osu.Game/Stores/RealmArchiveModelManager.cs

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Framework.Platform;
+using osu.Game.Database;
+using osu.Game.IO.Archives;
+using osu.Game.Models;
+using osu.Game.Overlays.Notifications;
+using Realms;
+
+#nullable enable
+
+namespace osu.Game.Stores
+{
+    /// <summary>
+    /// Class which adds all the missing pieces bridging the gap between <see cref="RealmArchiveModelImporter{TModel}"/> and <see cref="ArchiveModelManager{TModel,TFileModel}"/>.
+    /// </summary>
+    public abstract class RealmArchiveModelManager<TModel> : RealmArchiveModelImporter<TModel>, IModelManager<TModel>, IModelFileManager<TModel, RealmNamedFileUsage>
+        where TModel : RealmObject, IHasRealmFiles, IHasGuidPrimaryKey, ISoftDelete
+    {
+        public event Action<TModel>? ItemUpdated
+        {
+            // This may be brought back for beatmaps to ease integration.
+            // The eventual goal would be not requiring this and using realm subscriptions in its place.
+            add => throw new NotImplementedException();
+            remove => throw new NotImplementedException();
+        }
+
+        public event Action<TModel>? ItemRemoved
+        {
+            // This may be brought back for beatmaps to ease integration.
+            // The eventual goal would be not requiring this and using realm subscriptions in its place.
+            add => throw new NotImplementedException();
+            remove => throw new NotImplementedException();
+        }
+
+        private readonly RealmFileStore realmFileStore;
+
+        protected RealmArchiveModelManager(Storage storage, RealmContextFactory contextFactory)
+            : base(storage, contextFactory)
+        {
+            realmFileStore = new RealmFileStore(contextFactory, storage);
+        }
+
+        public void DeleteFile(TModel item, RealmNamedFileUsage file) =>
+            item.Realm.Write(() => DeleteFile(item, file, item.Realm));
+
+        public void ReplaceFile(TModel item, RealmNamedFileUsage file, Stream contents)
+            => item.Realm.Write(() => ReplaceFile(item, file, contents, item.Realm));
+
+        public void AddFile(TModel item, Stream stream, string filename)
+            => item.Realm.Write(() => AddFile(item, stream, filename, item.Realm));
+
+        public void DeleteFile(TModel item, RealmNamedFileUsage file, Realm realm)
+        {
+            item.Files.Remove(file);
+        }
+
+        public void ReplaceFile(TModel model, RealmNamedFileUsage file, Stream contents, Realm realm)
+        {
+            file.File = realmFileStore.Add(contents, realm);
+        }
```

Comment:
Somewhat perplexed by these.

* How do these differ from the overloads that go through `item.Realm`? Is it ever the case where the realm passed in would be different than `item.Realm`? Or, on the other side, why do the overloads without realms go through `item.Realm.Write()` rather than use the overloads with the realms?
* The unused params are a bit weird (`realm` on deletion, `model` on `ReplaceFile()`)

Reply:
signifies already in a transaction. will add xmldoc.

Reply:
I get the "already in transaction part", but I still don't get why the realm has to be passed in a separate argument rather than just using `item.Realm`. Is it ever valid to mutate a `RealmObject` on a realm other than `item.Realm`?

Reply:
I just added it as a marker to signify that a transaction is already open. The plan was that going forward, a `Realm` would always be passed into a function if the intention was that we are already in a transaction.

Generally (at least in realm documentation and examples) using `RealmObject.Realm` is not the correct way of doing things. We only use that in `RealmLive` because it saves on complexity, but I wouldn't recommend we do that anywhere else if we can instead pass the open realm.

Reply:
I'm kind of not sure why `RealmObject.Realm` is a property if its use is discouraged, but alright. It just irks me slightly that with the realms being passed in via argument there's a plausible scenario wherein a completely wrong realm is passed in, and it felt like using the above property would eliminate that.